File: blas/int32_arm64.go
package blas

import "unsafe"

//go:noescape

func _inc_int32(vec unsafe.Pointer, constant int32, size uint64)
func (o Int32s) Inc(constant int32) {
	_inc_int32(unsafe.Pointer(&o[0]), constant, uint64(len(o)))
}

func _add_int32(input1, input2, output unsafe.Pointer, size uint64)
func (o Int32s) Add(input1, input2 []int32) {
	_add_int32(unsafe.Pointer(&input1[0]), unsafe.Pointer(&input2[0]), unsafe.Pointer(&(o)[0]), uint64(len(o)))
}

func _sub_int32(input1, input2, output unsafe.Pointer, size uint64)
func (o Int32s) Sub(input1, input2 []int32) {
	_sub_int32(unsafe.Pointer(&input1[0]), unsafe.Pointer(&input2[0]), unsafe.Pointer(&(o)[0]), uint64(len(o)))
}

func _scalar_mul_int32(vec unsafe.Pointer, constant int32, size uint64)
func (o Int32s) ScalarMul(constant int32) {
	_scalar_mul_int32(unsafe.Pointer(&o[0]), constant, uint64(len(o)))
}

func _mul_int32(input1, input2, output unsafe.Pointer, size uint64)
func (o Int32s) Mul(input1, input2 []int32) {
	_mul_int32(unsafe.Pointer(&input1[0]), unsafe.Pointer(&input2[0]), unsafe.Pointer(&(o)[0]), uint64(len(o)))
}

func _inc_int32_sve(vec unsafe.Pointer, constant int32, size uint64)
func (o Int32s) IncSVE(constant int32) {
	_inc_int32_sve(unsafe.Pointer(&o[0]), constant, uint64(len(o)))
}

func _scalar_mul_int32_sve(vec unsafe.Pointer, constant int32, size uint64)
func (o Int32s) ScalarMulSVE(constant int32) {
	_scalar_mul_int32_sve(unsafe.Pointer(&o[0]), constant, uint64(len(o)))
}

File: blas/cpp/int32_arm64.cpp
#include <arm_neon.h>
#include <arm_sve.h>
#include <stdint.h>

void inc_int32(int32x4_t* vec, int32_t constant,  uint64_t size) {
    int chunks = size / 4;
    int32x4_t constant_vec = vdupq_n_s32(constant);

    for (int i = 0; i < chunks; i++) {
        vec[i] = vaddq_s32(vec[i], constant_vec);
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        ((int32_t *) vec)[i] = ((int32_t *) vec)[i] + constant;
    }
}

void add_int32(int32x4_t *input1, int32x4_t *input2, int32x4_t *output, uint64_t size) {
    int chunks = size / 4;

    for (int i = 0; i < chunks; i++) {
        output[i] = vaddq_s32(input1[i], input2[i]);
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        ((int32_t *) output)[i] = ((int32_t *) input1)[i] + ((int32_t *) input2)[i];
    }
}

void sub_int32(int32x4_t *input1, int32x4_t *input2, int32x4_t *output, uint64_t size) {
    int chunks = size / 4;

    for (int i = 0; i < chunks; i++) {
        output[i] = vsubq_s32(input1[i], input2[i]);
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        ((int32_t *) output)[i] = ((int32_t *) input1)[i] - ((int32_t *) input2)[i];
    }
}

void scalar_mul_int32(int32x4_t* vec, int32_t constant,  uint64_t size) {
    int chunks = size / 4;
    int32x4_t constant_vec = vdupq_n_s32(constant);

    for (int i = 0; i < chunks; i++) {
        vec[i] = vmulq_s32(vec[i], constant_vec);
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        ((int32_t *) vec)[i] = ((int32_t *) vec)[i] * constant;
    }
}

void mul_int32(int32x4_t *input1, int32x4_t *input2, int32x4_t *output, uint64_t size) {
    int chunks = size / 4;

    for (int i = 0; i < chunks; i++) {
        output[i] = vmulq_s32(input1[i], input2[i]);
    }
    // Leftovers
    for (int i = chunks * 4; i < size; i++) {
        ((int32_t *) output)[i] = ((int32_t *) input1)[i] * ((int32_t *) input2)[i];
    }
}


void inc_int32_sve(int32_t* vec, int32_t constant,  int size) {
    int i = 0;
    svint32_t constant_vec = svdup_s32(constant);

    while (i < size) {
        svbool_t pg = svwhilelt_b32(i, size);
        svst1(pg, &vec[i], svadd_s32_m(pg, svld1(pg, &vec[i]), constant_vec));
        i += svcntw();
    }
}

void scalar_mul_int32_sve(int32_t* vec, int32_t constant, int size) {
    int i = 0;
    svint32_t constant_vec = svdup_s32(constant);

    while (i < size) {
        svbool_t pg = svwhilelt_b32(i, size);
        svst1(pg, &vec[i], svmul_s32_m(pg, svld1(pg, &vec[i]), constant_vec));
        i += svcntw();
    }
}
